### PR TITLE
Fix Range’s len() + division and mod by 0

### LINKED ISSUE #331: Range should return actual length
In CPython, if you create a range with a positive step and a `start` larger than the `end`, then it will report zero length (as it actually produces no iterations). In RustPython it always returns the absolute value of `(end - start) / step`, which depending on the sign of the step might be wrong.

Compare this in CPython:
```
>>> r = range(10, 5)
>>> r
range(10, 5)
>>> print(len(r))
0
>>> for i in r: print(i)
... 
>>> 
```

with this in RustPython:
```
>>>>> r = range(10, 5)
>>>>> r
<range object at 0x5622737a1e28>
>>>>> print(len(r))
5
>>>>> for i in r: print(i)
>>>>> 
```

This also causes ranges to report wrong `len` when the difference between `start` and `end` is not divisible by `step`, eg.

In CPython:
```
>>> r = range(10, 5, -2)
>>> len(r)
3
>>> for i in r: print(i)
... 
10
8
6
```

while in RustPython:
```
>>>>> r = range(10, 5, -2)
>>>>> len(r)
2
>>>>> for i in r: print(i)
10
8
6
```

I believe the correct formula for the length is `ceil((end - start) / step)` with floating-point division.

CPython gets around this by special-casing positive and negative step, see its [get_len_of_range](https://github.com/python/cpython/blob/master/Objects/rangeobject.c#L834-L860) and [compute_range_length](https://github.com/python/cpython/blob/master/Objects/rangeobject.c#L144-L210)

Doing the same in RustPython would look something like:
```rust
#[inline]
pub fn len(&self) -> usize {
    match self.step.sign() {
        Sign::Plus if self.start < self.end =>
            ((&self.end - &self.start - 1usize) / &self.step).to_usize().unwrap() + 1,
        Sign::Minus if self.start > self.end =>
            ((&self.start - &self.end - 1usize) / (-&self.step)).to_usize().unwrap() + 1,
        _ => 0,
    }
}
```

### PR DESCRIPTION
**Pull request by silmeth**
Fixes #331 and #345.

### CODE REVIEW

**Comment by silmeth** [comment]
Now I see the same must be done with `divmod`, both for `int`s and for `float`s. Also, in CPython big integers (outside of `f64` range) are divisible if the result fits, eg.:

```
>>> 10**5000/10**5000
1.0
>>> 1/10**5000
0.0
```

I guess they use `divmod` internally to separately calculate the floor and the remainder of the division.

**Comment by orf** [comment]
Could we add a simple Python test case for this as well?

**Comment by silmeth** [comment]
I added a few test cases. Are they OK?

I guess one for `OverflowError` on the `len(range(…))` might also be a good idea.

What about `divmod`? It still panics on 0. I might try to fix it later, but not today and I might not have time to do it tomorrow. Would you consider merging this first as it is?

File: tests/snippets/division_by_zero.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+try:
+    5 / -0.0
+except ZeroDivisionError:
+    pass
+except:
+    assert False, 'Expected ZeroDivisionError'
+
+try:
+    5 / (3-2)
```
Comment (jmhain) [inline tests/snippets/division_by_zero.py]: Why is a `ZeroDivisionError` expected on this?

Comment (silmeth) [reply tests/snippets/division_by_zero.py]: It shouldn’t, it was supposed to be `(2-2)`… The fact that it passes at least caught a flaw in the tests themselves. Will fix them in a moment.

**Comment by windelbouwman** [comment]
Nice job!

**Pull request merged**